### ISSUE TITLE
Idle domain slows down major GC cycles

### ISSUE THREAD

**Issue by damiendoligez** [ocaml/ocaml#11589]
Consider the program below, which was reported to me by @Octachron. If you run it without options, it runs on a single domain without explicitly calling the GC and its heap size goes up to 31 MB. If you run it with `-spawn`, it launches a domain that only sleeps in parallel with its main computation, and its heap size goes up to 694 MB.

What happens is that the major GC slice budget is computed from the number of words allocated by the domain since the last slice. But the domain doesn't allocate and the slice budget is zero, so the slice doesn't do anything, even though the domain does have some work to do for the GC cycle (sweeping, for example). This ends up starving the GC cycle.

The GC still makes progress because of opportunistic slices, but these slices are very small and the GC runs a lot slower (6 cycles vs 116 cycles without the extra thread).

This is probably the same problem as #11548.

```ocaml
(* compile with:
  ocamlfind opt -package unix,domainslib -linkpkg sorttest.ml
*)
module Timings = struct

  type t = {
    cpu_time : float;
    wall_time : float
  };;

let time_computation (f : unit -> 'a) : t * 'a =
  let times_start = Sys.time ()
  and wall_start = Unix.gettimeofday () in
  let ret = f () in
  let times_end = Sys.time ()
  and wall_end = Unix.gettimeofday () in
  { cpu_time = times_end -. times_start ;
    wall_time = wall_end -. wall_start
  },
  ret
end
module Quicksort = struct
  let pivot (cmp : 'a -> 'a -> int) (a : 'a array) (left : int) (right : int) =
    assert(0 <= left);
    assert(left <= right);
    assert(right <= Array.length a);
    let mid_start = ref left
    and mid_end = ref (left + 1)
    and current_pos = ref (right - 1)
    and pivot = a.(left) in
    while !mid_end <= !current_pos do
      assert (left <= !mid_start);
      assert (!mid_start < !mid_end);
      assert (!current_pos < right);
      let current = a.(!current_pos) in
      match cmp current pivot with
      | 0 ->
        a.(!current_pos) <- a.(!mid_end);
        mid_end := !mid_end + 1
      | n when n < 0 ->
        a.(!mid_start) <- current;
        a.(!current_pos) <- a.(!mid_end);
        mid_start := !mid_start+1;
        mid_end := !mid_end+1
      | _ ->
        current_pos :=  !current_pos - 1
    done;
    for i= !mid_start to !mid_end-1 do
      a.(i) <- pivot
    done;
    (!mid_start, !mid_end);;

  let rec quicksort (cmp : 'a -> 'a -> int) (a : 'a array) (left : int) (right : int) =
    assert(0 <= left);
    assert(left <= right);
    assert(right <= Array.length a);
    if right - left > 1
    then
      let (mid_start, mid_end) = pivot cmp a left right in
      let len_left = mid_start - left
      and len_right = right - mid_end in
      if len_left < len_right then
        (quicksort cmp a left mid_start;
         quicksort cmp a mid_end right)
      else
        (quicksort cmp a mid_end right;
         quicksort cmp a left mid_start);;
end


[@@@warning "-unused-value-declaration"]


let assert_sorted cmp a left right =
  for i=left to right-2 do
    assert (cmp a.(i) a.(i+1) <= 0)
  done

let checksum a left right =
  let sum = ref 0 in
  for i=left to right-1 do
    sum := !sum lxor a.(i)
  done;
  !sum

let random_array seed max n =
  let prng = Random.State.make seed in
  Array.init n (fun _ -> Random.State.int prng max);; 

let print_array oc prt a left right =
  for i=left to right-1 do
    Printf.fprintf oc "[%d] = %a\n" i prt a.(i)
  done

let output_int oc n =
  output_string oc (string_of_int n);;

let pp_size ppf x =
  let x = x *. 8. in
  let rec human s suffixes x =
    match x<1024., suffixes with
    | true, _ | false, [] ->
      Printf.fprintf ppf "%.2f%sB" x s
    | false, s :: suffixes ->
      human s suffixes (x/.1024.)
  in
  human "" ["k";"M";"G";"T"] x
module Tsk = Domainslib.Task
(*  ;;
let cores = 4 in
let pool = Tsk.setup_pool ~num_additional_domains:(cores-1) () in
let run f = Tsk.run pool f in
*)
;;

type mode = Spawn_domains | Explicit_gc | Implicit_gc
let mode = ref Implicit_gc
let stride = ref 1000
let args = [
    "-spawn", Arg.Unit (fun () -> mode := Spawn_domains ), "spawn more than one domain";
    "-gc", Arg.Unit (fun () -> mode := Explicit_gc ), "add explicit call to the GC";
    "-default", Arg.Unit (fun () -> mode := Implicit_gc ), "no explicit call to the GC, single domain";
    "-stride", Arg.Set_int stride, "size increase at each step (default=1000)"
  ]

let () = Arg.parse args ignore ""
let mode = !mode
let stride = !stride
;;
let run f = f () in
let rec loop () =
  Unix.sleepf 100.0;
  loop ()
in
begin match mode with
| Spawn_domains ->
  let _d = Domain.spawn loop in
  Printf.eprintf "loop domain = %d\n%!" (Obj.magic (Domain.get_id _d))
| _ -> ()
end;
let n1 = ref 1 in
while !n1 <= 500_001 do
  let n = !n1 in
  n1 := n + stride;

  let max = 10000
  and seed = [|n; 2; 3|] in
  let a = random_array seed max n in
  let sum = checksum a 0 n in
  let times,() = Timings.time_computation (fun () ->
                     run
                     (fun () -> Quicksort.quicksort ( - ) a 0 n)) in
  assert_sorted ( - ) a 0 n ;
  let stat = Gc.quick_stat () in
  let sum2 = checksum a 0 n in
  let () = match mode with
    | Explicit_gc -> Gc.full_major ()
    | _ -> ()
  in
  assert (sum = sum2);
  Printf.printf "%d\t%f\t%f" n times.cpu_time times.wall_time;
  Printf.printf "\tmax=%a\theap=%a\tmajor allocated=%a minor=%d major=%d"
    pp_size (float stat.top_heap_words)
    pp_size (float stat.heap_words)
    pp_size stat.major_words
    stat.minor_collections
    stat.major_collections
  ;
  Printf.printf "\n";
  flush stdout
done;;
```

**Comment by damiendoligez** [comment]
I think we need to overhaul the GC slice computation to ensure that all domains are going through the GC work at the right pace to be on time for the global rendez-vous of the phase change.
Here is a possible design:
1. we maintain a global counter for allocations in the major heap
2. we know how much allocation will take place during this slice (from heap size and the overhead parameter)
3. from (1) and (2) we compute an allocation index (between 0 and 1) that reflects the progress of the slice
4. each domain computes (an approximation of) how much work it needs to do in the current cycle
5. from (3) and (4) each domain computes how much work it is supposed to have done by now
6. each domain sets its the slice sizes to meet the goal set by (5)

Most of this is re-use of the existing formulas for slice size computation.

For (1), I'm thinking of a global updated atomically at each slice.

What we still need to understand:
- Batched updates of the allocation pointer introduce a delay in the control loop. What is the effect of this delay?
- Some of the work (scanning of roots, ephemerons) is hard to predict and is not accounted for in the current design (nor in the sequential version). What is their effect?
- What to do when adopting orphaned work?
- Other things I'm overlooking.

cc @kayceesrk @stedolan @ctk21 @sadiqj

**Comment by kayceesrk** [comment]
@damiendoligez Thanks for looking into this. TBH we always knew that computing slice budget at each domain without a global view is sub-optimal, and this would be something that we would come back to after 5.0.

The above design seems like a reasonable start to me. One constraint that I would like to add is that, for sequential programs, the new slice budget computation either (a) attempts to match the slice budget in OCaml 4 or (b) computes a different budget but performs "better" than OCaml 4. 

> For (1), I'm thinking of a global updated atomically at each slice.

This sounds ok to me. 

> Batched updates of the allocation pointer introduce a delay in the control loop. What is the effect of this delay?

This looks unavoidable, but I would think that in practice this should be OK. We have a stop-the-world minor GC after which all the domains perform a major slice. Hence, there is no risk of a domain not participating in the new slice budget computation. 

FWIW, I have a PR in development for decoupling major slice for minor GC: https://github.com/kayceesrk/ocaml/pull/8. The PRs are orthogonal.

> What to do when adopting orphaned work?

We should ensure that when a domain terminates, its allocation counts is also orphaned so that it is accounted for by the domain which adopts it.

**Comment by EduardoRFS** [comment]
Any possible workaround besides of calling the GC by hand? A GC config to make it triggers more often would be ideal.

**Comment by sadiqj** [comment]
This seems like a good plan.

> We should ensure that when a domain terminates, its allocation counts is also orphaned so that it is accounted for by the domain which adopts it.

This deals with the allocations done by the terminated domain (for Damien's 3rd point), I think we might also need to account for the _work_ done by the terminated domain so the 4th point is accurate.

**Comment by damiendoligez** [comment]
@EduardoRFS 
>Any possible workaround besides of calling the GC by hand? A GC config to make it triggers more often would be ideal.

The GC is triggered often enough, the problem is in how it computes how much work to do when it is triggered. So I don't see any easy workaround besides explicitly calling `Gc.major_slice` with a non-zero argument that completely depends on your program.

**Comment by damiendoligez** [comment]
After brainstorming with @stedolan, here is the new design that I've started implementing:

- We count allocations globally.
- Allocations create an amount of budget that is distributed evenly among all domains.
- Each GC slice takes up its own domain's budget and does the corresponding work.
- When a domain has no more work to do in the current cycle, its overflow budget is redistributed to the other domains.
- When a domain terminates, its budget is also redistributed.
- When a new domain is spawned, it starts taking up its own share of the budget.


## Advantages:

Sweeping work is uneven between the domains: some have more sweeping to do, some have less. The ones that have less sweeping will start marking earlier, and thus do more marking (because they will grab more roots). This was part of the multicore design and we get to keep it.

We keep the open-loop control design of Ocaml 4: the GC makes progress at a constant rate with respect to allocations. This means we don't need to guess how much work will be needed per domain for the current GC cycle.

If there is only one domain, we get exactly the same behavior as in OCaml 4.

## Drawbacks / future work:

The load balancing for the marking work is coarse-grained. We will probably have to implement some kind of work-stealing algorithm in the future.

The GC pause latencies are not optimized. We'll have to see if they can be improved.

The distribution of work is fixed, we could probably be more clever (e.g by giving more work to idle domains).

**Comment by kayceesrk** [comment]
@damiendoligez the plan looks good. I don't see anything particularly problematic here. 

When you have the branch ready, please add them to Sandmark Nightly (https://discuss.ocaml.org/t/ann-sandmark-nightly-benchmarking-as-a-service/10174) so that we may see the nightly results for the sequential and parallel benchmarks on both of the parallel benchmarking machines -- `turing` and `navajo`. 

A few years ago, Stephen and I quickly played around with work-stealing in the multicore OCaml compiler. The idea was to steal half of the mark stack from the target domain, bound by a minimum threshold under which the work is not stolen. The early results were that stealing happened to be always slower. TBH we didn't put much work into optimising this, nor did we attempt to construct synthetic, best-case programs for stealing (a deep binary-tree, for example).
